src/app/storage/PascalString.h

```diff
@@ -0,0 +1,139 @@
+namespace Storage {
+
+/// Interprets a byte buffer as a pascal string with a 1-byte prefix.
+template <typename T>
+class ShortPascalString
+{
+public:
+    ShortPascalString(ShortPascalString &&)      = default;
+    ShortPascalString(const ShortPascalString &) = default;
+
+    template <size_t N>
+    ShortPascalString(T (&data)[N]) : mData(data), mMaxSize(N - 1)
+    {
+        static_assert(N > 0);
+        static_assert(N <= 0xFF);
+    }
+
+    /// Returns the content of the pascal string.
+    /// Uses the prefix size information
+    Span<T> Content() { return { mData + 1, GetLength() }; }
+    Span<const T> Content() const { return { mData + 1, GetLength() }; }
+
+    /// Accesses the "PASCAL" string (i.e. valid data includint the string perfix)
+    Span<const T> PascalContent() const { return { mData, static_cast<size_t>(GetLength() + 1) }; }
+
+    /// Access to the full buffer. does NOT take into account current size
+    /// and includes the "size prefix"
+    Span<T> Buffer() { return { mData, static_cast<size_t>(mMaxSize + 1) }; }
+
+    uint8_t GetLength() const { return std::min(mMaxSize, (reinterpret_cast<const uint8_t *>(mData))[0]); }
+    bool SetLength(size_t len)
+    {
+        VerifyOrReturnError(len <= mMaxSize, false);
+        mData[0] = static_cast<T>(len);
+        return true;
+    }
+
+    bool SetValue(Span<const T> value)
+    {
+        if (!SetLength(value.size()))
+        {
+            return false;
+        }
+        memcpy(mData + 1, value.data(), value.size());
+        return true;
+    }
+
+private:
+    T * mData;
+    const uint8_t mMaxSize;
+};
+
+/// Interprets a byte buffer as a pascal string with a 2-byte prefix.
+template <typename T>
+class LongPascalString
+{
+public:
+    LongPascalString(LongPascalString &&)      = default;
+    LongPascalString(const LongPascalString &) = default;
+
+    template <size_t N>
+    LongPascalString(T (&data)[N]) : mData(data), mMaxSize(N - 2)
+    {
+        static_assert(N > 0);
+        static_assert(N <= 0xFF);
+    }
+
+    /// Returns the content of the pascal string.
+    /// Uses the prefix size information
+    Span<T> Content() { return { mData + 2, GetLength() }; }
+    Span<const T> Content() const { return { mData + 2, GetLength() }; }
+
+    /// Accesses the "PASCAL" string (i.e. valid data includint the string perfix)
+    Span<const T> PascalContent() const { return { mData, static_cast<size_t>(GetLength() + 2) }; }
+
+    /// Access to the full buffer. does NOT take into account current size
+    /// and includes the "size prefix"
+    Span<T> Buffer() { return { mData, static_cast<size_t>(mMaxSize + 2) }; }
+
+    uint16_t GetLength() const
+    {
+        uint16_t len;
+        memcpy(&len, mData, 2);
+        return std::min(len, mMaxSize);
+    }
+
+    bool SetLength(size_t len)
+    {
+        VerifyOrReturnError(len <= mMaxSize, false);
+        auto lenBytes = static_cast<uint16_t>(len);
+        memcpy(mData, &lenBytes, 2);
```

Comment:
Note for future (if this PR is picked instead of re-making the Attribute storage): ember strings have a little endian prefix, so this needs fixing.

Reply:
Also need to treat 0xFF/0xFFFF as invalid strings